### ISSUE TITLE
Remove old (v0) titleindex (title pointer list)

### ISSUE THREAD

**Issue by kelson42** [openzim/libzim#800]
At the time of "removal" of namespaces, the "old" titleindex has been superseeded by the new "titleindex which is available as a normal item via [X/listing/titleordered/v1](https://wiki.openzim.org/wiki/Search_indexes#Title_index_v1).

We should considering removing the [old pointer list at ](https://wiki.openzim.org/wiki/Search_indexes#Title_index_v0). This should not be used anymore because deprecated 2 years ago and use storage "for nothing".

The move is easy just put `00` in the header in place on the title pointer list pointer and stop writing the old version of this pointer list.

But before doing that we should better check wit third party ZIM readers to verify that they don't use anymore:
* The old index
* The header pointer to the index

**Comment by Jaifroid** [comment]
Given the discussions on the complexities of supporting X/lisiting/titleOrdered/v1, and in particular on the fallbacks that are necessary if we don't find data in that listing (see https://github.com/kiwix/kiwix-js/issues/597 and https://github.com/kiwix/kiwix-js/issues/708), I would just urge a very cautious approach towards deprecating the `titlePtrList`, and a relatively long period of testing given the range of scrapers involved (i.e. different ZIM varieties).

**Comment by kelson42** [comment]
@Jaifroid I don't understand your comment: do we have any problem with the current ZIM spec and/or implementation on any of our software portfolio?

**Comment by Jaifroid** [comment]
@kelson42 No, but this is a low-level potentially breaking change, and because we still use a custom JS backend in Kiwix JS, I would like to be sure to have enough time to make any low-level adjustments if needed.

**Comment by kelson42** [comment]
@Jaifroid It's not breaking anything regarding current ZIM specification... if implementation of reader is OK. You are IMO the only one in position to check if kiwix JS implementation is OK and such a ticket won't break anything there. We need to be cautious with FUD effect.

**Comment by kelson42** [comment]
@Jaifroid Should I assume all good on your side and we could consider us ready for this issue version 10 of the libzim?

**Comment by Jaifroid** [comment]
@kelson42 Apologies, I'm on holiday so not responding as fast as normally. I can't say for sure that there would be no consequences without a ZIM to test. From memory, we use the `urlPtrPos` to find the listing for `X/titleOrdered/v1`, and fall back to `X/titleOrdered/v0`, then fall back to `titlePtrPos` in that order, so we should be OK. However, what I can say for sure is that if there are any issues, we can easily fix those with a test ZIM. So do go ahead.

**Comment by mgautierfr** [comment]
Will do this.

- Removing will be done only on the writing part.
- `titlePtrPos` in zim header will be filled with `0`
- Minor version will be updated. We don't need to bump major version as reader compatible with current major version should already read `X/titleOrdered/v1` first and so should be compatible with new version.
- Reading will still try to load `titlePtrPos` if it doesn't found `v1` version. (for compatibility with older zim files)
- Not being able to load neither `X/titleOrdered/v1` nor `v0`/`titlePtrPos` will be considered as hard error (invalid zim file) as it should never happen.

#### Testing:
New zim file without `titlePtrPos`/`v0` will be created and added to our test zim file (https://github.com/openzim/zim-testing-suite). Old zim files there will be keep.

**Comment by Jaifroid** [comment]
@mgautierfr Thanks, I'd be grateful if you could let me know when a test ZIM is available. The only part of the plan I need to check if I need to update support for is the minor version bump, but it would be an easy fix if an update is necessary in our backend.

**Comment by mgautierfr** [comment]
> The only part of the plan I need to check if I need to update support for is the minor version bump, but it would be an easy fix if an update is necessary in our backend.

I don't know how your reader is implemented but, if implemented as major/minor versions are intended to be used, you have nothing to do.

- Major version is a way to tell readers that if they don't know how to read version `M` they should not try (and stop as soon as possible)
- Minor version is a way to tell readers that new features has been added but they can ignore that and continue as zim file is compatible.

Say otherwise, a reader implemented with format/feature of zim version `M.N` (major.minor) when opening a zim:
- If major version of the zim is >M => Quit telling that it doesn't know how to read it.
- If minor version is >N => Continue as usual
- If minor version <N or major version <M => May (depending of implementation) detect that some feature will not be available in zim file and deactivate it. May also tell that zim file is too old and quit.

**Comment by Jaifroid** [comment]
@mgautierfr Thanks! I *think* it works like that. I did code that part, so can unpick any issues if I made wrong assumptions. I think there won't be a problem as I remember coding it to start with the X/... indices, falling back from one to the other, and only fall back to the pointer list if nothing else found. Since there should be no ZIM with missing X/ indices and a missing pointer list, I shouldn't have to handle that situation.

**Comment by kelson42** [comment]
@mgautierfr LGTM, don't forget to update spec wiki page.

**Comment by mgautierfr** [comment]
@Jaifroid, you can have some files without v0 title index from https://github.com/openzim/zim-testing-suite/pull/13

**Comment by kelson42** [comment]
@mgauthier I seems this has not been correctly implemented, see https://github.com/kiwix/kiwix-android-custom/issues/197#issuecomment-2786957798

**Comment by Jaifroid** [comment]
I somehow missed notification of ZIMs to test back in February. I have now tested `data/noTitleListingV0/wikibooks_be_all_nopic_2017-02.zim`, and while I get a warning in the PWA console `Unknown ZIM minor version: 3! Assuming content namespace is C`, the assumption is correct for that archive, and the landing page is correctly shown, navigation works.

I'll wait to see the outcome of this re-opened issue before adding explicit support for that minor version. (The warning is suppressed in the browser extension version.)

**Comment by mgautierfr** [comment]
One solution to fix this is to store 0xFFFFFFFFF in the header instead of 0x0.

Those changes https://github.com/openzim/libzim/pull/949/files#diff-7424944b2a718f9484b5f0bc782012673a1a2f2e9a8b2d8bf4663823d6f1e348L139 and https://github.com/openzim/libzim/pull/949/files#diff-808255119c8e0d757ab7cf077c5b8e50657d9888d35d2ea99ec540bb027d8bd9L314 would not be necessary and previous version of libzim would work.

If we postpone, we should at least add said changes to be prepare current libzim for the future change.

**Comment by mgautierfr** [comment]
> I'm sorry, but I don't understand what you mean. I'm happy to revert, but I fell with have to release version 10.0.0 right now (because of semantic versioning) if we include this fix. If this is not the case, glad to go forward with 9.2.4 or 9.3.0

I'm speaking only about the two linked changes.

Current libzim is still using `titleIdxPos` in the headers for two things:
- (Fast) check for data consistancy/corruption, we check that `titleIdxPos` > `mimeListPos`
- To find a upperbound limit to mimetype list to load in memory.

If we remove the titleIndexV0, we have to define something to put in `titleIdxPos`. In #949 we put zero (and we update libzim to no use `titleIdxPos` if it is equal to 0)
But older version are not update and still check `titleIdxPos`. Obviously the fast check fail with new zim file.

So either we put 0xFF..FF instead of 0 and we should be good, or we postpone the removing of titleindexv0, put we still should update the libzim to not check `titleIdxPos` when 0, to be prepare to the change in the format when we will actually release it.

**Comment by veloman-yunkan** [comment]
> So either we put 0xFF..FF instead of 0 and we should be good

It looks like we should be good with such a hack

**Comment by kelson42** [comment]
We have decided with @veloman-yunkan to make the foreseen fix and have this move part of release 9.3.0
